### ISSUE TITLE
[Feature Request] Sklearn example for Ray Tune

### ISSUE THREAD

**Issue by rohan-gt** [ray-project/ray#10163]
Is it possible to add a simple Sklearn example using the TuneBOHB algorithm with the HyperBandForBOHB scheduler in the documentation? 

P.S. This is the best hyperparameter tuning algorithm and scheduler right? If not an Sklearn example using the best combination of both would be very useful!

**Comment by richardliaw** [comment]
Thanks for opening this, @rohan-gt! What sklearn models are you using? IIRC most of them don't support iterative operations.

**Comment by rohan-gt** [comment]
All the different models actually. Also models having the Sklearn API like xgboost, lightgbm and catboost. Why is it necessary to have iterative operations? For early stopping?

**Comment by richardliaw** [comment]
Yeah; BOHB in Tune requires iterative operations.

**Comment by rohan-gt** [comment]
@richardliaw Is it possible to perform early stopping by cross-validation or using a sampled eval set from the train data?

**Comment by richardliaw** [comment]
@rohan-gt sorry for the slow reply. Can you tell me more about what you mean here?

**Comment by rohan-gt** [comment]
@richardliaw rather than split the entire train set into train and eval sets, perform early stopping using the cross-validated results on the entire train set. This is an early version of a method by Optuna for LightGBM: https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.lightgbm.LightGBMTunerCV.html

**Comment by richardliaw** [comment]
I’m interested in the second level: early stopping during cross validation. Do you mean like: I have 5 folds, and after 3 folds I realize this model is very bad, and therefore I stop training this model? 

Or do you mean something else?

**Comment by rohan-gt** [comment]
@richardliaw That was a separate feature that I suggested. This is different. Basically instead of feeding in separate train (80%) an eval (20%) sets for early stopping, the 100% train set can be input and early stopping can be done using the mean of the cross-validated results. This is how [lightgbm.cv](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.cv.html) works

**Comment by richardliaw** [comment]
OK, pardon my ignorance (I'm not that familiar with XGBoost)! I'll try to explain what you're suggesting, and please let me know if I am understanding you correctly:

```python
# Create the XGBoost model
clf = XGBoostClassifier()
# Create a Tune-sklearn object - use 4 separate CV splits
search = TuneSearchCV(clf, param_grid, cv=4)

# Only pass in the training data - but also stop the model when CV stops
search.fit(X_train, y_train, early_stopping_rounds=5)
```

Underneath the hood, it will do the following:
```
for parameter in param_grid:
	scores = []
	for x_train, y_train, x_test, y_test in kfold_split(X, y, k=4):
	    scores += [xgboost.train(x_train, y_train, parameter, early_stopping_rounds=5)]
	
	aggregate_scores(scores)
```
Is this the behavior that you expect? if so, I think this is already possible in tune-sklearn today. If I am mistaken, please let me know! 

Thank you for your patience!

**Comment by rohan-gt** [comment]
@richardliaw yes something similar. So now the early stopping parameter can be used for the 3 boosting algorithms?

**Comment by richardliaw** [comment]
So early_stopping_rounds + cv will work fine. However, tuner-level pruning will not work with all of them.

Does that make sense?

**Comment by rohan-gt** [comment]
@richardliaw yes it does

**Comment by richardliaw** [comment]
great @rohan-gt ! is there any outstanding feature requests you have re: tune or tune-sklearn?

**Comment by richardliaw** [comment]
OK closing this issue to track this on tune-sklearn side.
